### PR TITLE
Added Dataset Class, configured Query Class

### PR DESCRIPTION
**Pull request by aust427**
Added the ability to query by integer value for sample

Added the Dataset class including the following functions:
-convert (converts field by a scalar and appends to dataframe)
-aggs (similar to Query.aggs())
-view_pos_2D & view_pos_3D (for visualization)
-pdQuery (queries the local dataset)

Allowed functionality for a converted field to be queried in Query, aggs, and hist.

Added documentation for various functions

### CODE REVIEW

File: py/fi_astrosims/client.py [hunk 1]
```diff
@@ -55,36 +74,49 @@ def __init__(self, simulation, fields = None, sort = None, sample = 1, seed = No
         self.query = dict()
         if fields is None:
             fields = [ f['name'] for f in simulation.catalog['fields'] ]
```
Comment (dylex) [inline py/fi_astrosims/client.py]: This should be the original field name for transformed fields

File: py/fi_astrosims/client.py [hunk 2]
```diff
@@ -55,36 +74,49 @@ def __init__(self, simulation, fields = None, sort = None, sample = 1, seed = No
...
+        defaulting to all simulation fields.
+        :param sort: (type) specifies the ordering of result objects, which may
+        be the name of a single field, a field name prefixed with '-' for
+        descending, or a list of sort fields.
+        :param sample: (float) sets the fraction of results to return as a
+        random subset(0,1].
+        :param seed: (int) specifies the random seed for this selection, which
+        defaults to a different random set each time. If sample is set to a
+        value > 1, the value will be converted into a fraction within (0,1].
+        Note that each item is independently selected with the sample fraction,
```
Comment (dylex) [inline py/fi_astrosims/client.py]: Let's make this sentence part of the sample description.

File: py/fi_astrosims/client.py [hunk 3]
```diff
@@ -55,36 +74,49 @@ def __init__(self, simulation, fields = None, sort = None, sample = 1, seed = No
...
+        so the result set may not be exactly this fraction of the original size.
+        :param filters: (dict) parameters specify restrictions on fields to
+        query by. Each parameter must be the name of a field in the catalog,
+        and have the value of None to not filter on this field, a scalar to
+        filter exact values, or tuple of lower- and upper-bounds to select
+        items in that range.
+        Sample is applied first, then the remaining filters.
```
Comment (dylex) [inline py/fi_astrosims/client.py]: This probably shouldn't be the case, or should not be exposed to users.

File: py/fi_astrosims/client.py [hunk 4]
```diff
@@ -55,36 +74,49 @@ def __init__(self, simulation, fields = None, sort = None, sample = 1, seed = No
...
+        self._count = None
+        self._aggs = {}
 
-        Additional *filters* parameters specify restrictions on fields to query
-        by.  Each parameter must be the name of a field in the catalog, and
-        have the value of None to not filter on this field, a scalar to filter
-        exact values, or tuple of lower- and upper-bounds to select items in
-        that range.
-        """
+        self.count()
```
Comment (dylex) [inline py/fi_astrosims/client.py]: What's this count call doing? It'll cache it, but maybe should not. (count should be the count post-filtering)

File: py/fi_astrosims/client.py [hunk 5]
```diff
@@ -104,46 +136,151 @@ def update(self, fields = None, sort = None, sample = None, seed = None, **filte
...
-        self._aggs = {}
 
     def makeurl(self, path, **query):
         return makeurl(self.simulation.url, path, dictWith(self.query, **query))
 
     def count(self):
-        """Return the count of objects matching this query."""
+        """
+        :return: (int) Number of objects in the query
+        Can also obtain this value through aggs.
```
Comment (dylex) [inline py/fi_astrosims/client.py]: I think the count in aggs is for non-null fields, which may be different. Maybe not worth mentioning.

File: py/fi_astrosims/client.py [hunk 6]
```diff
@@ -104,46 +136,151 @@ def update(self, fields = None, sort = None, sample = None, seed = None, **filte
...
+        :param: field: (string) field you want to query
+        :return: (float) max value of that field
+        """
         return self.aggs(field)['max']
 
     def hist(self, field, width=None, bins=100):
-        """Return a histogram of the given field, with either the specified bin width or total bin count."""
+        """
+        :param field: (string) field that you want a histogram produced of
+        :param width: (int) width of each bin, either specified or calculated
```
Comment (dylex) [inline py/fi_astrosims/client.py]: width is really a float

File: py/fi_astrosims/client.py [hunk 7]
```diff
@@ -104,46 +136,151 @@ def update(self, fields = None, sort = None, sample = None, seed = None, **filte
...
+        :return: dataframe (pandas) with index-able fields 'bucket', 'count'
+        If you are querying on a transformed (scaled) field, scale factor will
+        adjust results accordingly.
+        """
+        scale = 1
+        if 'transform' in self.simulation.fields[field]:
+            scale = self.simulation.fields[field]['scale']
+            field = self.simulation.fields[field]['org-field']
         if not width:
             width = (self.max(field) - self.min(field))/bins
```
Comment (dylex) [inline py/fi_astrosims/client.py]: otherwise, width should have any transforms applied, no?

File: py/fi_astrosims/client.py (hunk 4 above)
Comment (aust427) [reply py/fi_astrosims/client.py]: If I remember correctly, this count call forces the count to be the number of items in the entire dataset.

**Pull request closed without merging**